[automation] memory leak caused by bad rule in UI

Tested against S2207 but reported on https://community.openhab.org/t/openhab-3-runs-out-of-memory-java-heap-space-errors-cpu-100-after-a-few-hours/110924/105 as happening between 3.0 and 3.0.1

As identified by @mhilbush a rule similar to the rule below, but created on the new rules UI.  This is 100% reproducible.  I created an Ubuntu 20.04 VM (on ESXi) with 4GB ram.  I only installed the systeminfo binding to monitor system use over time.  I removed the rrd4j persistence as it was thought to be causing the problem at once point (it wasn't).  There is nothing on this system other than a single thing created as per the config below.

```
rule "Test One per Second"
when
    Time cron "* * * * * ? *"
then
    xxxxxxx
    var Number n = 1
end
```

Causes the following:

```
2021-02-15 03:25:52.728 [DEBUG] [e.automation.internal.RuleEngineImpl] - The rule 'aeefe05726' is executed.
2021-02-15 03:25:53.673 [DEBUG] [e.automation.internal.RuleEngineImpl] - The trigger '1' of rule 'aeefe05726' is triggered.
2021-02-15 03:25:53.735 [ERROR] [internal.handler.ScriptActionHandler] - Script execution of rule with UID 'aeefe05726' failed:      ___ xxxxxxx
    var Number n = 1

   The method or field xxxxxxx is undefined; line 1, column 4, length 7
javax.script.ScriptException:      ___ xxxxxxx
    var Number n = 1

   The method or field xxxxxxx is undefined; line 1, column 4, length 7
        at org.openhab.core.model.script.runtime.internal.engine.DSLScriptEngine.eval(DSLScriptEngine.java:127) ~[?:?]
        at org.openhab.core.automation.module.script.internal.handler.ScriptActionHandler.lambda$0(ScriptActionHandler.java:62) [bundleFile:?]
        at java.util.Optional.ifPresent(Optional.java:183) [?:?]
        at org.openhab.core.automation.module.script.internal.handler.ScriptActionHandler.execute(ScriptActionHandler.java:59) [bundleFile:?]
        at org.openhab.core.automation.internal.RuleEngineImpl.executeActions(RuleEngineImpl.java:1179) [bundleFile:?]
        at org.openhab.core.automation.internal.RuleEngineImpl.runRule(RuleEngineImpl.java:987) [bundleFile:?]
        at org.openhab.core.automation.internal.TriggerHandlerCallbackImpl$TriggerData.run(TriggerHandlerCallbackImpl.java:89) [bundleFile:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]

```

Over time, this failure seems to cause a memory leak leading to an eventual system crash due to OutOfMemory

```
2021-02-15 02:22:34.305 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception:
java.lang.OutOfMemoryError: Java heap space
2021-02-15 02:22:34.307 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception:
java.lang.OutOfMemoryError: Java heap space
2021-02-15 02:22:34.307 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception:
java.lang.OutOfMemoryError: Java heap space
2021-02-15 02:22:34.307 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception:
java.lang.OutOfMemoryError: Java heap space
2021-02-15 02:22:34.306 [WARN ] [e.jetty.util.thread.QueuedThreadPool] -
java.lang.OutOfMemoryError: Java heap space
2021-02-15 02:22:34.308 [ERROR] [org.apache.felix.fileinstall        ] - In main loop, we have serious trouble
java.lang.OutOfMemoryError: Java heap space
2021-02-15 02:26:05.103 [WARN ] [e.jetty.util.thread.QueuedThreadPool] -
java.lang.OutOfMemoryError: Java heap space
2021-02-15 02:26:05.103 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception:
java.lang.OutOfMemoryError: Java heap space
```

systeminfo thing and items created as

```
openhab@oh3-test-vm:~$ cat /etc/openhab/things/systeminfo.things
Thing systeminfo:computer:test [interval_high=1, interval_medium=1]

openhab@oh3-test-vm:~$ cat /etc/openhab/items/systeminfo.items
String CPU_Name                   "Name"                <none>           { channel="systeminfo:computer:test:cpu#name" }
String CPU_Description            "Description"         <none>           { channel="systeminfo:computer:test:cpu#description" }
Number CPU_Load1                  "Load (1 min)"        <none>           { channel="systeminfo:computer:test:cpu#load1" }
Number CPU_Load5                  "Load (5 min)"        <none>           { channel="systeminfo:computer:test:cpu#load5" }
Number CPU_Load15                 "Load (15 min)"       <none>           { channel="systeminfo:computer:test:cpu#load15" }
Number CPU_Threads                "Threads"             <none>           { channel="systeminfo:computer:test:cpu#threads" }
Number CPU_Uptime                 "Uptime"              <time>           { channel="systeminfo:computer:test:cpu#uptime" }
Number Memory_Available           "Available"           <none>           { channel="systeminfo:computer:test:memory#available" }
Number Memory_Used                "Used"                <none>           { channel="systeminfo:computer:test:memory#used" }
Number Memory_Total               "Total"               <none>           { channel="systeminfo:computer:test:memory#total" }
Number Memory_Available_Percent   "Available (%)"       <none>           { channel="systeminfo:computer:test:memory#availablePercent" }
Number Memory_Used_Percent        "Used (%)"            <none>           { channel="systeminfo:computer:test:memory#usedPercent" }
```

I attempted to fix this in https://github.com/openhab/openhab-core/pull/2182 thinking that this was because of the threadpools but that was not successful.  I also attempted to back out https://github.com/openhab/openhab-core/pull/2057 and https://github.com/openhab/openhab-core/pull/1952 to see if either of those were the problem.  This was also fruitless.   I am at a loss on where to move further and can not identify where the leak is.

Text versus UI based evalutation and errors thrown

UI:

```
2021-02-15 19:54:47.355 [ERROR] [internal.handler.ScriptActionHandler] - Script execution of rule with UID 'aeefe05726' failed:      ___ xxxxxxx
    var Number n = 1

   The method or field xxxxxxx is undefined; line 1, column 4, length 7
javax.script.ScriptException:      ___ xxxxxxx
    var Number n = 1

   The method or field xxxxxxx is undefined; line 1, column 4, length 7
        at org.openhab.core.model.script.runtime.internal.engine.DSLScriptEngine.eval(DSLScriptEngine.java:128) ~[?:?]
        at org.openhab.core.automation.module.script.internal.handler.ScriptActionHandler.lambda$0(ScriptActionHandler.java:70) [bundleFile:?]
        at java.util.Optional.ifPresent(Optional.java:183) [?:?]
        at org.openhab.core.automation.module.script.internal.handler.ScriptActionHandler.execute(ScriptActionHandler.java:65) [bundleFile:?]
        at org.openhab.core.automation.internal.RuleEngineImpl.executeActions(RuleEngineImpl.java:1179) [bundleFile:?]
        at org.openhab.core.automation.internal.RuleEngineImpl.runRule(RuleEngineImpl.java:987) [bundleFile:?]
        at org.openhab.core.automation.internal.TriggerHandlerCallbackImpl$TriggerData.run(TriggerHandlerCallbackImpl.java:89) [bundleFile:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
```

Text:

```
2021-02-15 19:54:48.149 [ERROR] [internal.handler.ScriptActionHandler] - Script execution of rule with UID 'test2-1' failed: The name 'yyyyyyyy' cannot be resolved to an item or type; line 5, column 5, length 8 in test2
javax.script.ScriptException: The name 'yyyyyyyy' cannot be resolved to an item or type; line 5, column 5, length 8 in test2
        at org.openhab.core.model.script.runtime.internal.engine.DSLScriptEngine.eval(DSLScriptEngine.java:128) ~[?:?]
        at org.openhab.core.automation.module.script.internal.handler.ScriptActionHandler.lambda$0(ScriptActionHandler.java:70) [bundleFile:?]
        at java.util.Optional.ifPresent(Optional.java:183) [?:?]
        at org.openhab.core.automation.module.script.internal.handler.ScriptActionHandler.execute(ScriptActionHandler.java:65) [bundleFile:?]
        at org.openhab.core.automation.internal.RuleEngineImpl.executeActions(RuleEngineImpl.java:1179) [bundleFile:?]
        at org.openhab.core.automation.internal.RuleEngineImpl.runRule(RuleEngineImpl.java:987) [bundleFile:?]
        at org.openhab.core.automation.internal.TriggerHandlerCallbackImpl$TriggerData.run(TriggerHandlerCallbackImpl.java:89) [bundleFile:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]

```

--- a/bundles/org.openhab.core.model.script.runtime/src/org/openhab/core/model/script/runtime/internal/engine/ScriptEngineImpl.java
+++ b/bundles/org.openhab.core.model.script.runtime/src/org/openhab/core/model/script/runtime/internal/engine/ScriptEngineImpl.java
@@ -56,8 +56,6 @@ import org.slf4j.LoggerFactory;
 @Component(immediate = true)
 public class ScriptEngineImpl implements ScriptEngine, ModelParser {
 
-    protected XtextResourceSet resourceSet;
-
     private final Logger logger = LoggerFactory.getLogger(ScriptEngineImpl.class);
 
     private ScriptServiceUtil scriptServiceUtil;
@@ -72,16 +70,13 @@ public class ScriptEngineImpl implements ScriptEngine, ModelParser {
     }
 
     private XtextResourceSet getResourceSet() {
-        if (resourceSet == null) {
-            resourceSet = ScriptStandaloneSetup.getInjector().getInstance(XtextResourceSet.class);
-            resourceSet.addLoadOption(XtextResource.OPTION_RESOLVE_ALL, Boolean.FALSE);
-        }
+        XtextResourceSet resourceSet = ScriptStandaloneSetup.getInjector().getInstance(XtextResourceSet.class);
+        resourceSet.addLoadOption(XtextResource.OPTION_RESOLVE_ALL, Boolean.FALSE);
         return resourceSet;
     }
 
     @Deactivate
     public void deactivate() {
-        this.resourceSet = null;
         ScriptStandaloneSetup.unregister();
     }
 

--- a/bundles/org.openhab.core.model.script.runtime/src/org/openhab/core/model/script/runtime/internal/engine/ScriptEngineImpl.java
+++ b/bundles/org.openhab.core.model.script.runtime/src/org/openhab/core/model/script/runtime/internal/engine/ScriptEngineImpl.java
@@ -14,6 +14,7 @@ package org.openhab.core.model.script.runtime.internal.engine;
 
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
+import java.util.Collections;
 import java.util.List;
 import java.util.stream.Collectors;
 
@@ -56,6 +57,8 @@ import org.slf4j.LoggerFactory;
 @Component(immediate = true)
 public class ScriptEngineImpl implements ScriptEngine, ModelParser {
 
+    protected XtextResourceSet resourceSet;
+
     private final Logger logger = LoggerFactory.getLogger(ScriptEngineImpl.class);
 
     private ScriptServiceUtil scriptServiceUtil;
@@ -70,13 +73,16 @@ public class ScriptEngineImpl implements ScriptEngine, ModelParser {
     }
 
     private XtextResourceSet getResourceSet() {
-        XtextResourceSet resourceSet = ScriptStandaloneSetup.getInjector().getInstance(XtextResourceSet.class);
-        resourceSet.addLoadOption(XtextResource.OPTION_RESOLVE_ALL, Boolean.FALSE);
+        if (resourceSet == null) {
+            resourceSet = ScriptStandaloneSetup.getInjector().getInstance(XtextResourceSet.class);
+            resourceSet.addLoadOption(XtextResource.OPTION_RESOLVE_ALL, Boolean.FALSE);
+        }
         return resourceSet;
     }
 
     @Deactivate
     public void deactivate() {
+        this.resourceSet = null;
         ScriptStandaloneSetup.unregister();
     }
 
@@ -131,21 +137,23 @@ public class ScriptEngineImpl implements ScriptEngine, ModelParser {
 
         List<Diagnostic> errors = resource.getErrors();
         if (!errors.isEmpty()) {
+            deleteResource(resource);
             throw new ScriptParsingException("Failed to parse expression (due to managed SyntaxError/s)",
                     scriptAsString).addDiagnosticErrors(errors);
         }
 
         EList<EObject> contents = resource.getContents();
-
         if (!contents.isEmpty()) {
             Iterable<Issue> validationErrors = getValidationErrors(contents.get(0));
             if (!validationErrors.iterator().hasNext()) {
                 return (XExpression) contents.get(0);
             } else {
+                deleteResource(resource);
                 throw new ScriptParsingException("Failed to parse expression (due to managed ValidationError/s)",
                         scriptAsString).addValidationIssues(validationErrors);
             }
         } else {
+            deleteResource(resource);
             return null;
         }
     }
@@ -179,4 +187,11 @@ public class ScriptEngineImpl implements ScriptEngine, ModelParser {
         return "script";
     }
 
+    private void deleteResource(Resource resource) {
+        try {
+            resource.delete(Collections.emptyMap());
+        } catch(IOException e) {
+            // Do nothing
+        }
+    }
 }

